lib/models/plugin-registry.js

```diff
@@ -0,0 +1,205 @@
+      var name = self._pluginNameFromAddonName(currentAddon);
+
+      if (self._isValidDeployPlugin(currentAddon)) {
+        validAddons[name] = currentAddon;
+      }
+
+      if (self._isDeployPluginPack(currentAddon)) {
+        var nestedAddons = self._discoverValidAddons(currentAddon.addons);
+        validAddons = Object.keys(nestedAddons).reduce(function(v, key) {
+          v[key] = nestedAddons[key];
+          return v;
+        }, validAddons);
+      }
+
+      return validAddons;
+    }, {});
+  },
+
+  _buildAliasMap: function(addons, config) {
+    var self = this;
+    var aliasMap = Object.keys(addons)
+      .reduce(function(aliases, name) {
+        if (!aliases[name] || !aliases[name].as) {
+          aliases[name] = { as: [name] };
+        }
+
+        aliases[name].as = self._asArray(aliases[name], 'as');
+        return aliases;
+      }, config);
+
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(aliasMap), addons, 'config.pipeline.alias');
+
+    return aliasMap;
+  },
+
+  _buildDisabledMap: function(aliasMap, config, addons) {
+    var aliasArray = Object.keys(aliasMap)
+      .map(function(addonName) {
+        return aliasMap[addonName].as;
+      })
+      .reduce(function(aliases, arr) {
+        return aliases.concat(arr);
+      }, []);
+
+    var disabledMap = aliasArray.reduce(function(map, alias) {
+      if (map[alias] === undefined) {
+        map[alias] = false;
+      }
+
+      return map;
+    }, config);
+
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(disabledMap), addons, 'config.pipeline.disabled');
+
+    return disabledMap;
+  },
+
+  _buildRunOrderMap: function(config, addons) {
+    this._warnIfConfigReferencesUninstalledAddons(Object.keys(config), addons, 'config.pipeline.runOrder');
+
+    return config;
+  },
+
+  _installedPlugins: function(addons, aliasMap) {
+    return Object.keys(addons)
+      .map(function(addonName) {
+        var addon   = addons[addonName];
+        var aliases = aliasMap[addonName].as;
+
+        return aliases.map(function(alias) {
+          var v = addon.createDeployPlugin({ name: alias });
+          return v;
+        });
+      })
+      .reduce(function(plugins, arr) {
+        return plugins.concat(arr);
+      }, []);
+  },
+
+  _applyDisabledConfig: function(plugins, disabledMap) {
+    return plugins.filter(function(plugin) {
+      return !disabledMap[plugin.name];
+    }) ;
+  },
+
+  _applyRunOrderConfig: function(plugins, runOrderMap) {
+    var self          = this;
+    var graph         = new DAG();
+    var sortedPlugins = [];
+
+    try {
+      plugins.forEach(function(plugin) {
+        var before = self._asArray(runOrderMap[plugin.name], 'before').concat(self._asArray(plugin, 'runBefore'));
+        var after = self._asArray(runOrderMap[plugin.name], 'after').concat(self._asArray(plugin, 'runAfter'));
+        graph.addEdges(plugin.name, plugin, before, after);
+      });
```

Comment:
We previously considered ember addon DAG order. Do we intend to drop that altogether?